File: locus-api-core/src/main/java/locus/api/utils/Logger.java
package locus.api.utils;

public class Logger {

    private static ILogger logger;

    public static void registerLogger(ILogger logger) {
        Logger.logger = logger;
    }

    public static void logI(String tag, String msg) {
        if (logger == null) {
            System.out.println(tag + " - " + msg);
        } else {
            logger.logI(tag, msg);
        }
    }

    public static void logD(String tag, String msg) {
        if (logger == null) {
            System.out.println(tag + " - " + msg);
        } else {
            logger.logD(tag, msg);
        }
    }

    public static void logW(String tag, String msg) {
        if (logger == null) {
            System.out.println(tag + " - " + msg);
        } else {
            logger.logW(tag, msg);
        }
    }

    public static void logE(String tag, String msg) {
        if (logger == null) {
            System.err.println(tag + " - " + msg);
        } else {
            logger.logE(tag, msg);
        }
    }

    public static void logE(String tag, String msg, Exception e) {
        if (logger == null) {
            System.err.println(tag + " - " + msg + ", e:" + e.getMessage());
            e.printStackTrace();
        } else {
            logger.logE(tag, msg, e);
        }
    }

    public interface ILogger {

        void logI(String tag, String msg);

        void logD(String tag, String msg);

        void logW(String tag, String msg);

        void logE(String tag, String msg);

        void logE(String tag, String msg, Exception e);
    }
}

File: locus-api-core/src/main/java/locus/api/utils/Logger.kt
package locus.api.utils

/**
 * Generic static logger instance used in Locus API system.
 * To receive correct log messages, initialize over `registerLogger` as soon as possible.
 */
object Logger {

    private var logger: ILogger? = null

    fun registerLogger(logger: ILogger) {
        Logger.logger = logger
    }

    fun logI(tag: String, msg: String) {
        logger?.logI(tag, msg)
                ?: println("$tag - $msg")
    }

    fun logD(tag: String, msg: String) {
        logger?.logD(tag, msg)
                ?: println("$tag - $msg")
    }

    fun logW(tag: String, msg: String) {
        logger?.logW(tag, msg)
                ?: println("$tag - $msg")
    }

    fun logE(tag: String, msg: String) {
        logger?.logE(tag, msg)
                ?: System.err.println("$tag - $msg")
    }

    fun logE(tag: String, msg: String, e: Exception) {
        logger?.logE(tag, msg, e)
                ?: {
                    System.err.println(tag + " - " + msg + ", e:" + e.message)
                    e.printStackTrace()
                }()
    }

    interface ILogger {

        fun logI(tag: String, msg: String)

        fun logD(tag: String, msg: String)

        fun logW(tag: String, msg: String)

        fun logE(tag: String, msg: String)

        fun logE(tag: String, msg: String, e: Exception)
    }
}
